### PR TITLE
TEST/IODEMO: Pass memory handle hint to send/recv functions

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Test user-defined memory handle with io_demo

## How
- Port object pool refactoring from master branch: needed to avoid passing ucp context to all pools
- Port user memo from master branch #7761

### CODE REVIEW

**Comment by yosefe** [comment]
@evgeny-leksikov @gleon99 can you pls review?

File: test/apps/iodemo/io_demo.cc [hunk 1]
```diff
@@ -145,13 +150,39 @@ class MemoryPool {
...
+    }
 
+    void fill_offcache_queue()
+    {
+        while (_offcache_queue.size() < _num_offcache) {
+            _offcache_queue.push(get_new());
+        }
+    }
+
+     inline BufferType* get_free()
```
Comment (gleon99) [inline test/apps/iodemo/io_demo.cc]: Formatting (indent)

Comment (gleon99) [inline test/apps/iodemo/io_demo.cc]: `inline BufferType *get_free()`

File: test/apps/iodemo/io_demo.cc [hunk 2]
```diff
@@ -145,13 +150,39 @@ class MemoryPool {
...
+     {
+        BufferType* item;
```
Comment (gleon99) [inline test/apps/iodemo/io_demo.cc]: `BufferType *item;`

File: test/apps/iodemo/io_demo.cc [hunk 3]
```diff
@@ -145,13 +150,39 @@ class MemoryPool {
...
         if (_free_stack.empty()) {
-            item = new T(_buffer_size, *this);
-            _num_allocated++;
+            // Fill the offcache queue on first use. Assume the free stack will
+            // also be empty on the first use.
+            if (_offcache_queue.size() < _num_offcache) {
```
Comment (gleon99) [inline test/apps/iodemo/io_demo.cc]: This condition is already checked inside `fill_offcache_queue()`.

File: test/apps/iodemo/io_demo.cc [hunk 4]
```diff
@@ -94,18 +95,22 @@ typedef struct {
...
-    MemoryPool(size_t buffer_size, const std::string& name, size_t offcache = 0) :
-        _num_allocated(0), _buffer_size(buffer_size), _name(name) {
-
-        for (size_t i = 0; i < offcache; ++i) {
-            _offcache_queue.push(get_free());
+    ObjectPool(size_t buffer_size, size_t num_offcache,
+               const std::string &name) :
+        _buffer_size(buffer_size),
+        _num_offcache(num_offcache),
+        _num_allocated(0),
```
Comment (gleon99) [inline test/apps/iodemo/io_demo.cc:105]: Minor, not related to the change.
Maybe move to `_num_allocated = 0;` at the bottom of the class?
Ctor initialization means value depends on arg, which is not true here.

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -372,7 +465,9 @@ class P2pDemoCommon : public UcxContext {
     private:
         void*                     _buffer;
         size_t                    _size;
-        MemoryPool<Buffer, true>& _pool;
+        BufferMemoryPool<Buffer> *_pool;
```
Comment (gleon99) [inline test/apps/iodemo/io_demo.cc]: Can avoid that? (Replacing ref with a ptr).

File: test/apps/iodemo/io_demo.cc (hunk 4 above)
Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:105]: @gleon99 could you pls explain why
> Ctor initialization means value depends on arg

and why
> `_num_allocated = 0;`

is better for ctor?

Comment (evgeny-leksikov) [reply test/apps/iodemo/io_demo.cc:105]: never mind, I see what you mean - c++11 default initialization inside of class definition

Comment (yosefe) [reply test/apps/iodemo/io_demo.cc:105]: Currently we don't build iodemo with c++11..

File: test/apps/iodemo/io_demo.cc (hunk 5 above)
Comment (yosefe) [reply test/apps/iodemo/io_demo.cc]: fixed

**Comment by yosefe** [comment]
@gleon99 your comments were applied, can you pls re-review?

**Review by gleon99 — APPROVED** [review]
👍

**Review by evgeny-leksikov — APPROVED** [review]

**Pull request merged**